storage/pebble/registers_pruner.go

```diff
@@ -0,0 +1,326 @@
+// ticker interval.
+func WithPruneTickerInterval(interval time.Duration) PrunerOption {
+	return func(p *RegisterPruner) {
+		p.pruneTickerInterval = interval
+	}
+}
+
+// WithPrunerMetrics is used to sets the metrics for a RegisterPruner instance.
+func WithPrunerMetrics(metrics module.RegisterDBPrunerMetrics) PrunerOption {
+	return func(p *RegisterPruner) {
+		p.metrics = metrics
+	}
+}
+
+func NewRegisterPruner(
+	logger zerolog.Logger,
+	db *pebble.DB,
+	opts ...PrunerOption,
+) (*RegisterPruner, error) {
+	pruner := &RegisterPruner{
+		logger:              logger.With().Str("component", "register_pruner").Logger(),
+		db:                  db,
+		pruneInterval:       pruneInterval(DefaultPruneThreshold),
+		pruneThreshold:      DefaultPruneThreshold,
+		pruneThrottleDelay:  DefaultPruneThrottleDelay,
+		pruneTickerInterval: DefaultPruneTickerInterval,
+	}
+
+	pruner.componentManager = component.NewComponentManagerBuilder().
+		AddWorker(pruner.loop).
+		Build()
+	pruner.Component = pruner.componentManager
+
+	for _, opt := range opts {
+		opt(pruner)
+	}
+
+	return pruner, nil
+}
+
+// loop is the main worker for the Pruner, responsible for triggering
+// pruning operations at regular intervals. It monitors the heights
+// of registered height recorders and checks if pruning is necessary.
+func (p *RegisterPruner) loop(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+	ticker := time.NewTicker(p.pruneTickerInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			p.checkPrune(ctx)
+		}
+	}
+}
+
+// checkPrune checks if pruning should be performed based on the height range
+// and triggers the pruning operation if necessary.
+//
+// Parameters:
+//   - ctx: The context for managing the pruning operation.
+func (p *RegisterPruner) checkPrune(ctx irrecoverable.SignalerContext) {
+	// TODO: Clarify whether batching can be used here for getting the first and latest heights,
+	//       and for updating the first height.
+	firstHeight, err := firstStoredHeight(p.db)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to get first height from register storage: %w", err))
+	}
+
+	latestHeight, err := latestStoredHeight(p.db)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to get latest height from register storage: %w", err))
+	}
+
+	pruneHeight := latestHeight - p.pruneThreshold
+
+	if pruneHeight-firstHeight > p.pruneInterval {
+		p.logger.Info().Uint64("prune_height", pruneHeight).Msg("pruning storage")
+
+		// update first indexed height
+		err = p.updateFirstStoredHeight(pruneHeight)
+		if err != nil {
+			ctx.Throw(fmt.Errorf("failed to update first height for register storage: %w", err))
+		}
+
+		err := p.pruneUpToHeight(pruneHeight)
+		if err != nil {
+			ctx.Throw(fmt.Errorf("failed to prune: %w", err))
+		}
+
+		if p.metrics != nil {
+			p.metrics.NumberOfBlocksPruned(pruneHeight - firstHeight)
+		}
```

Comment:
by convention, `metrics` should always be non-nil. If we want to not record metrics, pass a noop collector instead
```suggestion
		p.metrics.NumberOfBlocksPruned(pruneHeight - firstHeight)
```

Reply:
[25e1356](https://github.com/onflow/flow-go/pull/6397/commits/25e1356776fae69a82abceed975f470cc300798d)